RdlMigration/ConvertRDL.cs

```diff
@@ -45,32 +57,41 @@ public void ConvertFolder(string urlEndPoint, string inputPath, string workspace
 
             if (!rdlFileIO.IsFolder(inputPath))
             {
-                ConvertAndUploadReport(
-                                        powerBIPortal,
-                                        rdlFileIO,
-                                        inputPath);
+                rootFolder = Path.GetDirectoryName(inputPath).Replace("\\", "/");
+                var reportName = Path.GetFileName(inputPath);
+                reportNameMap.TryAdd(reportName, 1);
+                reportPaths.Add(inputPath);
             }
             else
             {
-                var reportPaths = rdlFileIO.GetReportsInFolder(inputPath);
-
-                Console.WriteLine($"Found {reportPaths.Length} reports to convert");
-                Parallel.ForEach(reportPaths, reportPath => ConvertAndUploadReport(
-                                                                                powerBIPortal,
-                                                                                rdlFileIO,
-                                                                                reportPath));
+                rootFolder = inputPath;
+                var rootReports = rdlFileIO.GetReportsInFolder(inputPath);
+                Console.WriteLine($"Found {rootReports.Length} reports to convert");
+                foreach (string reportPath in reportPaths)
+                {
+                    var reportName = Path.GetFileName(reportPath);
+                    reportNameMap.TryAdd(reportName, 1);
+                }
+                reportPaths.AddRange(rootReports);
+            }
+            foreach (string reportPath in reportPaths.ToList())
```

Comment:
Arent you going to miss subreports in the conversion? E.g you have report1 that uses subreport1 which uses subreport2. In DiscoverSubreport you are adding back the reports to reportPaths, but since you called ToList here before you startedProcessing (since you can change the list under the enumerable anyway) the new files that you want to add to the list would not be processed. Instead why don't you use an actual queue in a BFS fashion and process elements until your queue is empty?

Reply:
fixed using queue instead of list